### ISSUE TITLE
Move all packages to ESM

### ISSUE THREAD

**Issue by domoritz** [vega/vega#3990]
We are planning to move all Vega packages to ESM only. Other packages like D3 have done this a while ago and it reduces future maintenance burden for us.

This will require a major release for all packages.

- [x] Vega https://github.com/vega/vega/pull/4018
- [x] Vega-Lite https://github.com/vega/vega-lite/pull/9527
- [x] Vega Embed https://github.com/vega/vega-embed/pull/1384
- [x] Vega Tooltip https://github.com/vega/vega-tooltip/pull/1000
- [x] Vega Themes https://github.com/vega/vega-themes/pull/651
- [x] Vega Datasets https://github.com/vega/vega-datasets/pull/673
- [ ] ~~ts-json-schema-generator~~

See https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c and https://github.com/d3/d3/issues/3469.

**Comment by hydrosquall** [comment]
> it reduces future maintenance burden for us.

This sounds good to me, It could be beneficial for future maintainers reading about this move for us to to see which of these burdens are heaviest right now. The Sindresorhus link covers how, and the D3 post talks about node.js incompatibility, but perhaps we can have our own short recap of the why (especially given that given its interactive roots, I'd predict most vega users are in a browser rather than server-side rendering). 

Do you have a recommended article o list of highlights for what we gain or lose by moving to ESM? 

My understanding of the main consequences might include that we

1. Drop support for older browsers (pre 2015) using CJS  as we assume usage in this group is low 
2. Don't have to ship compiled JS in the repo anymore (i.e. https://github.com/vega/vega/blob/main/docs/vega.min.js ), making releasing changes easier (we can just edit the source files)
3. Align w/ JS ecosystem as a whole is moving towards ESM only (at 26% as of Nov 2024)
4. Get general ESM benefits: modern tools rely on ESM for developer benefits like tree-shaking, hot-reloading, etc per Anthony's [post](https://antfu.me/posts/move-on-to-esm-only)
5. Fewer complications when it comes to making proper types for exported objects when providing Typescript types 
6. Fewer complications when it comes to resolving dependencies w/ conflicting versions ( e.g.  #3932 )
7. Smaller package sizes (by not having to ship both versions)

**Comment by domoritz** [comment]
Yeah, having the es5 support for too long meant we had to test more configurations. For now, I still kept the umd bundles but maybe we should remove them which would allow us more flexibility in bundlers. I plan to look more into that after doing a first pass. 

CJS is annoying in that it makes tracking dependencies harder (bad for bundlers like vite) and many tools are moving to esm formats for configs (eslint, etc) and so we need a mix of cjs and esm files in repos. Moving to `type: module` should unify and simplify things (e.g. https://github.com/vega/vega-embed/pull/1366). Having everything in esm and fewer output formats should also make it easier to switch to bundlers like deno, bun, esbuild, etc.

**Comment by domoritz** [comment]
I'm removing the umd bundles in favor of esm.

**Comment by domoritz** [comment]
Done!
